Makers Academy - Week 1 Weekend Challenge

PKilgarriff opened the pull request:
# Paul Kilgarriff

Please write your full name here to make it easier to find your pull request.

# User stories 

Please list which user stories you've implemented (delete the ones that don't apply).

- [x] User story 1: "I want to instruct a plane to land at an airport"
- [x] User story 2: "I want to instruct a plane to take off from an airport and confirm that it is no longer in the airport"
- [x] User story 3: "I want to prevent landing when the airport is full"
- [x] User story 4: "I would like a default airport capacity that can be overridden as appropriate"
- [x] User story 5: "I want to prevent takeoff when weather is stormy"
- [x] User story 6: "I want to prevent landing when weather is stormy"

# README checklist

Does your README contains instructions for

- [x] how to install,
- [x] how to run,
- [x] and how to test your code?

[Here is a pill](https://github.com/makersacademy/course/blob/main/pills/readmes.md) that can help you write a great README!

PKilgarriff committed:
Create plane_spec
 - add landing test
 - add take off test
 - add plane leaves after take off test

PKilgarriff committed:
Add Plane class
Add Airport class

PKilgarriff committed:
Add trailing empty lines

PKilgarriff committed:
Add tests to plane_spec
 - tests for airborne
 - test for registration information

PKilgarriff committed:
Add tests to airport_spec
 - test for weather update
 - test for full airport
 - test for plane landing
 - test for plane take off

PKilgarriff committed:
Update Plane class
 - add initialize method
 - add airborne? method
 - add registration method
 - remove land method
 - remove take_off method

PKilgarriff committed:
Update Airport class
 - add current_weather instance variable
 - add land method with exception raising
 - add take_off method with exception raising
 - add weather_update method
 - add full? method
 - add stormy? method

PKilgarriff committed:
Update spec files
 - plane_spec: remove output line in registration test
 - rearrange airport_spec into instance method groups

PKilgarriff committed:
Add name instance variable to Airport class

PKilgarriff committed:
Add attr_reader for make and serial_number

PKilgarriff committed:
Update plane_spec
 - Add test for change_location

PKilgarriff committed:
Update take_off method in Airport class
 - take serial number as argument
 - call change_location within method

PKilgarriff committed:
Add change_location method to Plane class

PKilgarriff committed:
Update airport spec and Airport
 - test that an error is raised when taking off an unavailable plane
 - raise error if plane cannot be found

PKilgarriff committed:
Rename variables with numbers to be normalcase

PKilgarriff committed:
Update airport_spec and plane_spec
 - instance double of plane for airport_spec
 - plane and airport created as lets for DRY

PKilgarriff committed:
Update Airport class
 - use constant to set default capacity
 - add fetch_plane method

PKilgarriff committed:
Update Plane class & spec
 - change_location: correct comparison to assign (== -> =)
 - remove unnecessary respond_to test
 - update airborne? test to check for correct return

PKilgarriff committed:
Refactor Plane and Airport
 - remove unnecessary explicit returns
 - remove commented code
 - change {} to do...end in spec

PKilgarriff committed:
Update Airport class
 - remove report_capacity in favour of capacity attr_reader
 - add change list_planes to print_planes
 - move methods to private
   - full?
   - stormy?
   - fetch_plane

PKilgarriff committed:
Update airport_spec
 - remove tests for methods that are now private
 - update tests to account for new method names
 - add test for taking off in stormy weather

PKilgarriff committed:
Add has_plane? method

PKilgarriff committed:
Add feature_spec
 - add model for multiple airports and planes

PKilgarriff committed:
Add Weather Class
 - add class method weather_report
 - add constant for weather conditions

PKilgarriff committed:
Update Airport class
 - requires Weather class
 - has_plane? renamed to contains_plane?
 - stormy? now uses weather_report

PKilgarriff committed:
Update airport_spec
 - add class_double for Weather
 - add test for print_planes
 - change references from weather_update to update_weather
 - change references from has_plane? to contains_plane?

PKilgarriff committed:
Update Plane class
 - remove change_location method in favour of attr_writer
 - update spec accordingly

PKilgarriff committed:
Update Weather class
 - change weather_report from class to instance method
 - use rand(1..20) instead of sample from array

PKilgarriff committed:
Update Airport Class
 - use Plane class's attr_writer for location
 - use Weather class's weather_report method

PKilgarriff committed:
Rubocop tweaks
 - trailing newline
 - spaces around = signs

PKilgarriff committed:
Change specs to use weather double
 - instance_double instead of class_double
 - use allow for specific weather conditions in tests

PKilgarriff committed:
Add weather_spec.rb
 - add test to check weather_report

PKilgarriff committed:
Update AIrport and Plane
 - remove airport require for weather
 - remove commented method in plane

PKilgarriff committed:
Add feature test
 - multiple planes take off to land in different airports

PKilgarriff committed:
Create Challenge README
 - renamed original README to Challenge-Brief.md
 - Provided preliminary overview of approach
 - included basic user guide
 - list of next steps for project

PKilgarriff committed:
Corrected typo

PKilgarriff committed:
Update Readme
 - include running code
 - include testing code

jonnyabrams commented on spec/airport_spec.rb:
```diff
@@ -0,0 +1,81 @@
+  let(:airport) { Airport.new("Heathrow") }
+  let(:plane) { instance_double("Plane", :serial_number => serial_number) }
+  let(:weather) { instance_double(Weather, :weather_report => "Sunny") }
+
```
These let clauses look proper snazzy - gonna have to read up on this

PKilgarriff replied:
Highly recommend - have a resource that I will link that is very overwhelming, but gives you places to jump off from.
https://relishapp.com/rspec/rspec-mocks/docs

PKilgarriff replied:
Above was not the deer in the headlights resource - this one is: https://github.com/rspec/rspec-mocks#method-stubs
Useful but needs chipping away at

jonnyabrams commented on lib/plane.rb:
```diff
@@ -0,0 +1,18 @@
+    @location = "Unknown"
+    @make = make
+    @serial_number = serial_number
+  end
```
The creation of real planes with serial numbers is a great touch

PKilgarriff replied:
Thanks - definitely a distraction from writing the code, but kept me happier

jonnyabrams reviewed (commented):
Here y'go

valentina-maggio commented on lib/airport.rb:
```diff
@@ -0,0 +1,54 @@
+class Airport
+  DEFAULT_CAPACITY = 10
+  attr_reader :capacity
```
If you use the attr_reader, can the capacity be changed by the user?

PKilgarriff replied:
No not currently - it's only able to be overridden at the point of creation. Do you think I should add that in as a method?

valentina-maggio replied:
I thought using the attr_accessor would best suit for the purpose of letting the the user change the capacity any time while running the program. I used the walkthrough example from the Boris Bike challenge as a guideline for this: 
https://github.com/makersacademy/course/blob/main/boris_bikes/walkthroughs/17.md

PKilgarriff replied:
Thank you, I'll take a look. My only worry with that would be that a user could change it to anything, not just an integer

valentina-maggio replied:
Good point! I didn't think about that!

valentina-maggio commented on lib/airport.rb:
```diff
@@ -0,0 +1,54 @@
+  end
+
+  def land(plane)
+    raise "Airport is full" if full?
```
I would move the 'raise' bits of code in a separate method to respect the SRP.

PKilgarriff replied:
Absolutely, I will borrow your example!

valentina-maggio commented on lib/airport.rb:
```diff
@@ -0,0 +1,54 @@
+  def stormy?
+    @weather == "Stormy"
+  end
+end
```
Code very clear and readable!

valentina-maggio commented on spec/airport_spec.rb:
```diff
@@ -0,0 +1,81 @@
+      end.to raise_error 'Cannot take off during stormy weather'
+    end
+  end
+end
```
I suggest to use 'context' based on the weather if it's stormy or not

PKilgarriff replied:
Yes will do - need to read up on the usage

PKilgarriff replied:
Will follow your example and use contexts inside the instance methods

PKilgarriff committed:
Update after Code Review
 - group tests under contexts, e.g. stormy weather
 - use named subject instead of let in spec files
 - remove commented-out code

PKilgarriff committed:
Add travis config file
